File: ch-5/quickSort.py
def partition(arr, low, high):
    i = (low - 1)
    piv = arr[high]

    for j in range(low, high):
        if arr[j] <= piv:
            i = i + 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quickSort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quickSort(arr, low, pi - 1)
        quickSort(arr, pi + 1, high)


arr = []
n = int(input("Enter the number of element in array"))
for i in range(0, n):
    ele = int(input(f"Enter the {i} element: "))
    arr.append(ele)

quickSort(arr,0,n-1)
print("\nSorted array using QuickSort:",end=" ")
for i in range(n):
    print(arr[i],end=" ")

File: ch-5/quickSort.cpp
#include<bits/stdc++.h>
using namespace std;
void swap(int *a, int *b){
	int temp;
	temp = *a;
	*a = *b;
	*b = temp;
}

int Partition(int a[], int low, int high){
	int pivot, index, i;
	index = low;
	pivot = high;

	for(i=low; i < high; i++){
		if(a[i] < a[pivot]){
			swap(&a[i], &a[index]);
			index++;
		}
	}
	swap(&a[pivot], &a[index]);
	return index;
}

int RandomPartition(int a[], int low, int high){
	int pvt, n, temp;
	n = rand();
	pvt = low + n%(high-low+1);
	swap(&a[high], &a[pvt]);
	return Partition(a, low, high);
}

int QuickSort(int a[], int low, int high){
	int pindex;
	if(low < high){
		pindex = RandomPartition(a, low, high);
		QuickSort(a, low, pindex-1);
		QuickSort(a, pindex+1, high);
	}
	return 0;
}

void display(int arr[],int n){
    cout<<endl<<"Sorted Array using QuickSort is: ";
	for (int i=0;i<n;i++){
        cout<<arr[i]<<" ";
	}
}

int main(){
	int m, i;
	cout<<"Enter the number elements in the array: ";
	cin>>m;

	int arr[m];
	for(i = 0; i < m; i++){
		cout<<"Enter element "<<i+1<<": ";
		cin>>arr[i];
	}

	QuickSort(arr, 0, m-1);
	display(arr,m);
	return 0;
}
